File: backend/src/main/java/com/todoapp/service/CategoryService.java
package com.todoapp.service;

import com.todoapp.dto.CategoryDTO;
import com.todoapp.dto.CreateCategoryRequest;
import com.todoapp.dto.UpdateCategoryRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for category operations.
 */
public interface CategoryService {
    
    /**
     * Get all categories for a user.
     * @param userEmail User email
     * @return List of categories
     */
    List<CategoryDTO> getCategories(String userEmail);
    
    /**
     * Get all categories for a user (without pagination).
     * @param userEmail User email
     * @return List of categories
     */
    List<CategoryDTO> getAllCategories(String userEmail);
    
    /**
     * Create a new category.
     * @param request Category creation request
     * @param userEmail User email
     * @return Created category
     */
    CategoryDTO createCategory(CreateCategoryRequest request, String userEmail);
    
    /**
     * Update a category.
     * @param id Category ID
     * @param request Category update request
     * @param userEmail User email
     * @return Updated category
     */
    CategoryDTO updateCategory(Long id, UpdateCategoryRequest request, String userEmail);
    
    /**
     * Delete a category.
     * @param id Category ID
     * @param userEmail User email
     */
    void deleteCategory(Long id, String userEmail);
    
    /**
     * Get a specific category.
     * @param id Category ID
     * @param userEmail User email
     * @return Category details
     */
    CategoryDTO getCategory(Long id, String userEmail);
} 

File: frontend/src/services/categoryService.ts
import { apiService } from './api';
import type { Category, CreateCategoryRequest, UpdateCategoryRequest } from '../types';

export const categoryService = {
  // Get all categories
  getCategories: async (): Promise<Category[]> => {
    const response = await apiService.get<Category[]>('/categories');
    return response;
  },

  // Get category by ID
  getCategory: async (id: number): Promise<Category> => {
    const response = await apiService.get<Category>(`/categories/${id}`);
    return response;
  },

  // Create new category
  createCategory: async (categoryData: CreateCategoryRequest): Promise<Category> => {
    const response = await apiService.post<Category>('/categories', categoryData);
    return response;
  },

  // Update category
  updateCategory: async (id: number, categoryData: UpdateCategoryRequest): Promise<Category> => {
    const response = await apiService.put<Category>(`/categories/${id}`, categoryData);
    return response;
  },

  // Delete category
  deleteCategory: async (id: number): Promise<void> => {
    await apiService.delete(`/categories/${id}`);
  },
}; 